### PR TITLE
[PSL] Better Logging and Option to Skip Validation + Truncate for PSL Migration

### PR DESCRIPTION
**Pull request by gmaretic**
**Goals (and why)**:
This PR does two things: improves the logging on the migration and validation tasks, and adds a timelock install config that allows us to skip validation and truncate old logs when we need this to move forward if safe with failing validation

**Implementation Description (bullets)**:

- Added `toString()` to `PaxosAcceptorState`, basically no reason not to do it this way
- Wired in and added client and use case to the logs
- Added a `truncateAllRounds` method to PSL interface
- If the config is set to true, skip validation if migrated, and in any case truncate the file based log
Funnily enough found two bugs in PaxosStateLogImpl so fixed them along the way.

**Testing (What was existing testing like?  What have you done to improve it?)**:
Basically added some new tests that are modified existing tests and verify that the changed behaviour makes sense.

**Concerns (what feedback would you like?)**:
Not the biggest fan of adding more dangerous stuff to the interface, but the alternative was to change the behaviour of `PaxosStateLogImpl` to truncate everything. The JavaDoc implicitly allows this, but in the vein of being more defensive we just don't need this kind of implementation detail change to somehow kick is in the teeth.

**Where should we start reviewing?**:
Tests

**Priority (whenever / two weeks / yesterday)**:
🔥

### CODE REVIEW

File: leader-election-impl/src/main/java/com/palantir/paxos/PaxosStateLogImpl.java [hunk 1]
```diff
@@ -223,7 +224,7 @@ public void truncate(long toDeleteInclusive) {
             for (File file : files) {
                 long fileSeq = getSeqFromFilename(file);
                 if (fileSeq <= toDeleteInclusive) {
-                    if (file.delete()) {
+                    if (!file.delete()) {
                         log.warn("failed to delete log file {}", file.getAbsolutePath());
```
Comment (jeremyk-91) [inline leader-election-impl/src/main/java/com/palantir/paxos/PaxosStateLogImpl.java:228]: wat

Comment (jeremyk-91) [reply leader-election-impl/src/main/java/com/palantir/paxos/PaxosStateLogImpl.java:228]: good spot though!

File: leader-election-impl/src/main/java/com/palantir/paxos/PaxosStateLogImpl.java [hunk 2]
```diff
@@ -235,10 +236,24 @@ public void truncate(long toDeleteInclusive) {
...
     }
 
+    @Override
+    public void truncateAllRounds() {
+        lock.writeLock().lock();
+        try {
+            File dir = new File(path);
+            boolean success = getLogEntries(dir).stream().map(File::delete).reduce(true, (x, y) -> x && y);
+            if (!success) {
+                log.warn("failed to delete all log files.");
```
Comment (jeremyk-91) [inline leader-election-impl/src/main/java/com/palantir/paxos/PaxosStateLogImpl.java:246]: nit: could we include, at least at `Unsafe` level, the directory involved?

File: leader-election-impl/src/main/java/com/palantir/paxos/PaxosStateLogImpl.java [hunk 3]
```diff
@@ -235,10 +236,24 @@ public void truncate(long toDeleteInclusive) {
...
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
     private List<File> getLogEntries(File dir) {
         File[] files = dir.listFiles();
         if (files == null) {
-            return null;
+            return ImmutableList.of();
```
Comment (jeremyk-91) [inline leader-election-impl/src/main/java/com/palantir/paxos/PaxosStateLogImpl.java:256]: I realise this case is handled already above in `getExtremeLogEntry`, but we should clean it up.

**Review by jeremyk-91 — APPROVED** [review]
🕐 0:32, 🏁 2:29

Implementation looks good, apart from a little thing in PaxosStateLogImpl that should be cleaned up. Thanks for very detailed tests!

My main question here is whether we really want to couple both "skip the check" and "truncate the old log". I imagine we might want to do the first to unblock a stack which we know to not appear corrupted but not the second if we want to try and do an RCA. But I think the name is clear enough, and someone on storage support should know that truncating the log means, well, truncating the log.
